### PR TITLE
support swagger property annotation to extend description properties of model

### PR DESCRIPTION
**Pull request by kufd**
Hi
I implemented new functionality which can help to make model definition better
I allow to use swagger property annotation to add or redefine some properties of model options
For example you can add description, title or redefine type

My changes do not change previous behavior but add more improvements

Also i changes ObjectModelDescriber now it works similar to JMSModelDescriber
Because in way how it was workin before it was not possible to add for example propertiy descriptions (all simple types referenced to the same definition)

And i updated tests to cover changes too

### CODE REVIEW

**Comment by GuilhemN** [comment]
Thank you for submitting this PR! I think I'll go with https://github.com/nelmio/NelmioApiDocBundle/pull/1119 which does less changes for a similar result but that's for the time you took to do this.

**Comment by kufd** [comment]
@GuilhemN 
Thank you for your reply
But i would like to add my approach is very different from  #1119
I enreach property information using swagger annotation but not rewrite it completely. It is more flexible and convenient
Here swagger annotation works with default and JMS object describers
Also its is very inconveniently when default object describer use definitions for scalar fields (int, string), but it is changed in my pull request

All my changes pass tests

And when to do major changes if not in beta version, especially when behavior is not changed

File: ModelDescriber/ObjectModelDescriber.php [hunk 1]
```diff
@@ -54,12 +79,51 @@ public function describe(Model $model, Schema $schema)
...
+                        $this->modelRegistry->register(new Model($type, $model->getGroups()))
+                    );
+                }
+            } else {
+                throw new \Exception("Unknow type: {$type->getBuiltinType()}");
+            }
+
+            //read property options from Swagger Property annotation if it exists
+            if (property_exists($class, $propertyName)) {
+                $property = $this->swaggerPropertyAnnotationReader->readSwaggerPropertyAnnotation(
```
Comment (GuilhemN) [inline ModelDescriber/ObjectModelDescriber.php]: No need to store the returned value

File: ModelDescriber/ObjectModelDescriber.php [hunk 2]
```diff
@@ -18,17 +18,42 @@
 use Symfony\Component\PropertyInfo\PropertyInfoExtractorInterface;
 use Symfony\Component\PropertyInfo\Type;
 
+/**
+ * Class ObjectModelDescriber
+ * @package Nelmio\ApiDocBundle\ModelDescriber
```
Comment (GuilhemN) [inline ModelDescriber/ObjectModelDescriber.php]: No need of this :)

File: .gitignore [hunk 3]
```diff
@@ -7,3 +7,4 @@
 /.phpunit
 /Tests/Functional/cache
 /Tests/Functional/logs
+/.idea
```
Comment (GuilhemN) [inline .gitignore]: Should be removed?

File: ModelDescriber/ObjectModelDescriber.php [hunk 4]
```diff
@@ -54,12 +79,51 @@ public function describe(Model $model, Schema $schema)
...
+                if (in_array($type->getClassName(), ['DateTime', 'DateTimeImmutable'])) {
+                    $property->setType('string');
+                    $property->setFormat('date-time');
+                } else {
+                    $property->setRef(
+                        $this->modelRegistry->register(new Model($type, $model->getGroups()))
+                    );
+                }
+            } else {
+                throw new \Exception("Unknow type: {$type->getBuiltinType()}");
```
Comment (GuilhemN) [inline ModelDescriber/ObjectModelDescriber.php]: Please use `sprintf` instead

File: ModelDescriber/SwaggerPropertyAnnotationReader.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Nelmio\ApiDocBundle\ModelDescriber;
+
+use EXSyst\Component\Swagger\Schema;
+use EXSyst\Component\Swagger\Items;
+use Swagger\Annotations\Property as SwgProperty;
+use Doctrine\Common\Annotations\Reader as AnnotationsReader;
```
Comment (GuilhemN) [inline ModelDescriber/SwaggerPropertyAnnotationReader.php]: No need to alias it

File: ModelDescriber/SwaggerPropertyAnnotationReader.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+     */
+    public function __construct(AnnotationsReader $annotationsReader)
+    {
+        $this->annotationsReader = $annotationsReader;
+    }
+
+    /**
+     * @param \ReflectionProperty $reflectionProperty
+     * @param Items|Schema        $property
+     * @return Items|Schema
```
[the remark this answers is not in the document]
Comment (kufd) [reply ModelDescriber/SwaggerPropertyAnnotationReader.php]: Thank you!
I will fix them this evening

**Comment by kufd** [comment]
Thank you!
I will fix them this evening

File: ModelDescriber/ObjectModelDescriber.php (hunk 1 above)
Comment (kufd) [reply ModelDescriber/ObjectModelDescriber.php]: It is rather way to show that property is modified inside readSwaggerPropertyAnnotation method

File: ModelDescriber/SwaggerPropertyAnnotationReader.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (kufd) [reply ModelDescriber/SwaggerPropertyAnnotationReader.php]: I think it is better to keep this dock block
I cannot use type hint for parameter $property, because Items and Schema classes do not implement common interface. And all they method are from traits.

And i think it is usefull to see what types of parameter $property are expected.

File: ModelDescriber/ObjectModelDescriber.php (hunk 1 above)
Comment (GuilhemN) [reply ModelDescriber/ObjectModelDescriber.php]: Then maybe the method should not be prefixed with `read` ;) What about `updateWith...`?

File: ModelDescriber/SwaggerPropertyAnnotationReader.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (GuilhemN) [reply ModelDescriber/SwaggerPropertyAnnotationReader.php]: Ok this block wasn't the best example, but most of the others can be removed

**Comment by kufd** [comment]
I made fixes for all comments to PR, execpt one
But there is an issue with test environmen config, so all tests are failed

**Comment by kufd** [comment]
Now all should be ok

**Comment by kufd** [comment]
Thank you

File: ModelDescriber/JMSModelDescriber.php [hunk 7]
```diff
@@ -91,6 +99,9 @@ public function describe(Model $model, Schema $schema)
                     $this->modelRegistry->register(new Model(new Type(Type::BUILTIN_TYPE_OBJECT, false, $type), $model->getGroups()))
                 );
             }
+
+            //read property options from Swagger Property annotation if it exists
+            $property = $this->swaggerPropertyAnnotationReader->updateWithSwaggerPropertyAnnotation($item->reflection, $property);
```
Comment (GuilhemN) [inline ModelDescriber/JMSModelDescriber.php]: No need to save the returned value ;)

File: ModelDescriber/JMSModelDescriber.php [hunk 8]
```diff
@@ -91,6 +99,9 @@ public function describe(Model $model, Schema $schema)
                     $this->modelRegistry->register(new Model(new Type(Type::BUILTIN_TYPE_OBJECT, false, $type), $model->getGroups()))
                 );
             }
+
+            //read property options from Swagger Property annotation if it exists
```
Comment (GuilhemN) [inline ModelDescriber/JMSModelDescriber.php]: Space missing

File: ModelDescriber/ObjectModelDescriber.php [hunk 9]
```diff
@@ -54,9 +60,44 @@ public function describe(Model $model, Schema $schema)
...
+                if (in_array($type->getClassName(), ['DateTime', 'DateTimeImmutable'])) {
+                    $property->setType('string');
+                    $property->setFormat('date-time');
+                } else {
+                    $property->setRef(
+                        $this->modelRegistry->register(new Model($type, $model->getGroups()))
+                    );
+                }
+            } else {
+                throw new \Exception(sprintf("Unknow type: %s", $type->getBuiltinType()));
```
Comment (GuilhemN) [inline ModelDescriber/ObjectModelDescriber.php]: Unknown

File: ModelDescriber/SwaggerPropertyAnnotationReader.php [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            }
+            if ($swgProperty->title !== null) {
+                $property->setTitle($swgProperty->title);
+            }
+            if ($swgProperty->example !== null) {
+                $property->setExample((string) $swgProperty->example);
+            }
+        }
+
+        return $property;
```
Comment (GuilhemN) [inline ModelDescriber/SwaggerPropertyAnnotationReader.php]: No need to return the value

File: ModelDescriber/SwaggerPropertyAnnotationReader.php [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * @param \ReflectionProperty $reflectionProperty
+     * @param Items|Schema        $property
+     * @return Items|Schema
+     */
+    public function updateWithSwaggerPropertyAnnotation(\ReflectionProperty $reflectionProperty, $property)
+    {
+        $swgProperty = $this->annotationsReader->getPropertyAnnotation($reflectionProperty, SwgProperty::class);
+        if ($swgProperty instanceof SwgProperty) {
```
Comment (GuilhemN) [inline ModelDescriber/SwaggerPropertyAnnotationReader.php]: What about rather using an early return for readability? :)

Comment (kufd) [reply ModelDescriber/SwaggerPropertyAnnotationReader.php]: Honestly, i do not like this approach)
But if it is important i will change.

**Comment by kufd** [comment]
I have done fixes for comments
Also some tests were failed because symfony 4 has bacome stable and it was issue with jms serializer
So i have done that symfony 3 is installed by default

File: .travis.yml [hunk 12]
```diff
@@ -27,6 +27,7 @@ matrix:
 
 before_install:
     - phpenv config-rm xdebug.ini
+    - composer require symfony/symfony:^3.0 --dev --no-update;
```
Comment (GuilhemN) [inline .travis.yml]: Please revert this change, sf 4 is installés on purpose.

Comment (kufd) [reply .travis.yml]: Done

**Comment by GuilhemN** [comment]
Thank you @kufd!

**Comment by kufd** [comment]
You are welcome @GuilhemN 
And If any issues regarded to my changes will be found i am ready to fix them in sort time
